### PR TITLE
ssh: Use libssh's knownhost functions 

### PR DESCRIPTION
**Pull request by Gundersanne**

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -132,6 +132,13 @@ if test "$enable_ssh" != "no"; then
     AC_DEFINE_UNQUOTED(HAVE_SSH_GET_SERVER_PUBLICKEY, 1, Whether ssh_get_server_publickey is available)
   ])
 
+  AC_CHECK_LIB(ssh, ssh_session_has_known_hosts_entry, [
+    has_session_known_host_entry="yes"
+  ])
+  AC_CHECK_LIB(ssh, ssh_session_has_known_hosts_entry, [
+    AC_DEFINE_UNQUOTED(HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY, 1, Whether ssh_has_known_host_entry is available)
+  ])
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply configure.ac]: Right, I'll do that, did't konw you could have multiple actions after a check_lib

File: src/ssh/cockpitsshknownhosts.c [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "common/cockpitknownhosts.h"
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshknownhosts.c]: Can't see why not, I'll do that.

File: src/ssh/cockpitsshrelay.c [hunk 3]
```diff
@@ -591,16 +538,33 @@ prepend_host_port (const gchar *host,
...
+static gboolean
+session_has_known_host_in_file (const gchar *file,
+                               CockpitSshData *data,
+                               const gchar *host,
+                               const guint port)
+{
+#ifndef HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
+    shim_set_knownhosts_file(file);
+    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_HOST, host) == SSH_OK);
+    g_warn_if_fail (ssh_options_set (data->session, SSH_OPTIONS_PORT, &port) == SSH_OK);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Oh i think those two set's are redundant actually, very much a WIP :)

File: src/ssh/cockpitsshrelay.c [hunk 4]
```diff
@@ -610,14 +574,17 @@ set_knownhosts_file (CockpitSshData *data,
...
+  /* check the default global ssh file */
+  if (!host_known)
+      host_known = session_has_known_host_in_file (data->ssh_options->knownhosts_file, data, host, port);
 
   /* if we check the default system known hosts file (i. e. not during the test suite), also check
    * the legacy file in /var/lib/cockpit */
   if (!host_known && strcmp (data->ssh_options->knownhosts_file, cockpit_get_default_knownhosts ()) == 0)
     {
-      host_known = cockpit_is_host_known (LEGACY_KNOWN_HOSTS, host, port);
+      host_known = session_has_known_host_in_file (LEGACY_KNOWN_HOSTS, data, host, port);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: I can include it here.

File: src/ssh/cockpitsshrelay.c [hunk 5]
```diff
@@ -667,12 +634,11 @@ set_knownhosts_file (CockpitSshData *data,
...
                 goto out;
 
-              /* this should always be known now, but let's make double-sure */
-              host_known = cockpit_is_host_known (tmp_knownhost_file, host, port);
+              host_known = session_has_known_host_in_file (tmp_knownhost_file, data, host, port);
               if (host_known)
                 data->ssh_options->knownhosts_file = tmp_knownhost_file;
               else
-                g_warning ("sss_ssh_knownhostsproxy reported key for %s:%u which is not known to cockpit_is_host_known()", host, port);
+                g_debug ("sss_ssh_knownhostsproxy reported key for %s:%u which is not known to session_has_known_host_in_file()", host, port);
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/cockpitsshrelay.c]: Yup, which is why I temporarily changed it, as the sssd_badkey one does fail. 

The libssh api just recognizes it as unknown at this stage.

However the old cockpit api for some reason doesn't see the bad key as unknown, but as known; proceeds to connect but fails because the key is invalid.

I think this is one of the only instances where code path actually changes, and in this case I think it's totally fine/makes more sense.

File: src/ssh/test-sshbridge.c [hunk 6]
```diff
@@ -606,8 +606,11 @@ test_echo_large (TestCase *tc,
   json_object_unref (init);
 }
 
-
+#ifdef HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
+#define MOCK_RSA_KEY "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCYzo07OA0H6f7orVun9nIVjGYrkf8AuPDScqWGzlKpAqSipoQ9oY/mwONwIOu4uhKh7FTQCq5p+NaOJ6+Q4z++xBzSOLFseKX+zyLxgNG28jnF06WSmrMsSfvPdNuZKt9rZcQFKn9fRNa8oixa+RsqEEVEvTYhGtRf7w2wsV49xIoIza/bln1ABX1YLaCByZow+dK3ZlHn/UU0r4ewpAIZhve4vCvAsMe5+6KJH8ft/OKXXQY06h6jCythLV4h18gY/sYosOa+/4XgpmBiE7fDeFRKVjP3mvkxMpxce+ckOFae2+aJu51h513S9kxY2PmKaV/JU9HBYO+yO4j+j24v\n"
+#else
 #define MOCK_RSA_KEY "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCYzo07OA0H6f7orVun9nIVjGYrkf8AuPDScqWGzlKpAqSipoQ9oY/mwONwIOu4uhKh7FTQCq5p+NaOJ6+Q4z++xBzSOLFseKX+zyLxgNG28jnF06WSmrMsSfvPdNuZKt9rZcQFKn9fRNa8oixa+RsqEEVEvTYhGtRf7w2wsV49xIoIza/bln1ABX1YLaCByZow+dK3ZlHn/UU0r4ewpAIZhve4vCvAsMe5+6KJH8ft/OKXXQY06h6jCythLV4h18gY/sYosOa+/4XgpmBiE7fDeFRKVjP3mvkxMpxce+ckOFae2+aJu51h513S9kxY2PmKaV/JU9HBYO+yO4j+j24v"
+#endif
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/test-sshbridge.c]: We can! I'll add the linebreak :)

File: src/ssh/test-sshbridge.c [hunk 7]
```diff
@@ -747,7 +750,11 @@ static const TestFixture fixture_knownhost_sssd_badkey = {
   .knownhosts_sssd = "ssh-rsa Ym9ndXMK",
   .knownhosts_sssd_host = "127.0.0.1",
   .knownhosts_file = "/dev/null",
+#ifdef HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
+  .problem = "unknown-host"
+#else
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply src/ssh/test-sshbridge.c]: See previous comment. Not entirely sure what to do in this situation as libssh api recognizes it as unknown because it's invalid while the cockpit api gets through the pre-connection check.

**Comment by Gundersanne** [comment]
@pitti The second paragraph isn't super clear, sorry for that. `this` means adding `ssh_has_known_host_entry` to the shim as I did with `ssh_session_export_known_hosts_entry`.

It just means there's a small ifdef that calls `shim_set_knownhosts_file` in the relay.

File: configure.ac [hunk 8]
```diff
@@ -132,6 +132,11 @@ if test "$enable_ssh" != "no"; then
     AC_DEFINE_UNQUOTED(HAVE_SSH_GET_SERVER_PUBLICKEY, 1, Whether ssh_get_server_publickey is available)
   ])
 
+  AC_CHECK_LIB(ssh, ssh_session_has_known_hosts_entry, [
+      AC_DEFINE_UNQUOTED(HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY, 1, Whether ssh_has_known_host_entry is available)
+    has_session_known_host_entry="yes"
```
Comment (martinpitt) [inline configure.ac]: OK if that works, but it looks rather unusual, and the indentation is off, too. Why isn't the `AM_CONDITIONAL` below enough to define (or not) `HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY`?

Also, please have a look at my PR #10241. `AC_CHECK_DECLS` is much simpler to use.

Comment (Gundersanne) [reply configure.ac]: Ah thanks, I'll take a look. To be honest I don't have that much experience with automake.

Comment (Gundersanne) [reply configure.ac]: Also I need both a  preprocessor variable and a variable inside the Makefile. Can that be achieved with just `AM_CONDITIONAL`?

Comment (martinpitt) [reply configure.ac]: `AC_CHECK_DECLS` automatically defines an `ac_cv_have_decl_ssh_session_has_known_hosts_entry` shell variable in this case, which you can use with `AM_CONDITIONAL` afterwards.

**Comment by martinpitt** [comment]
Can you please split out the dropping of the legacy known_hosts file into a separate commit?

**Comment by Gundersanne** [comment]
@martinpitt  I pushed with a big comment/commit message. I think sadly we simply can't offer this information (which exact file the (invalid) key was found in and the user should/could inspect). Not with the libssh api at least. 

So I'd see this as a limitation (not being able to ask which specific file the key was found in). Haven't thought of a good solution yet.

File: src/ssh/Makefile-ssh.am [hunk 9]
```diff
@@ -63,6 +70,17 @@ test_sshbridge_CFLAGS = $(cockpit_ssh_CFLAGS)
...
+SSH_CHECKS += \
+	test-knownhosts \
+	$(NULL)
+
+test_knownhosts_CFLAGS = $(cockpit_ssh_CFLAGS)
+test_knownhosts_SOURCES = src/ssh/test-knownhosts.c
+test_knownhosts_LDADD = libcockpit-ssh.a $(cockpit_ssh_LDADD)
+endif
+
+
```
Comment (Gundersanne) [inline src/ssh/Makefile-ssh.am]: 1 newline too many

**Comment by martinpitt** [comment]
If not knowing which particular file has an offending key is the only obstacle here, then I think we shouldn't try to work around this if it takes a nontrivial amount of effort. Let's report this as a bug against libssh, that `ssh_session_has_known_hosts_entry()` does not return the file on failure, and not show it. 

ISTM that our current logic to find the known_hosts file is too complicated. Naïvely, this ought to be enough:

 * If we got a host key from the protocol, or through `$COCKPIT_SSH_KNOWN_HOSTS_FILE`, set this as known_hosts location. 
 * Otherwise, the internal defaults of libssh should be as one expects already -- check global (/etc/ssh/ssh_known_hosts) and user (~/.ssh/known_hosts) files. Ideally we wouldn't need to call anything there, or perhaps just `ssh_options_set(SSH_OPTIONS_{,GLOBAL_}KNOWNHOSTS, NULL)` once.
 * If the host isn't known by the above, ask sssd, and if it has an answer, write that into a temp file, set it as GLOBAL_KNOWN_HOSTS, and retry `ssh_session_is_known_server()`.

Would that work?

**Comment by martinpitt** [comment]
```
ERROR: files left in build directory after distclean:
./core
./libpreload-temp-home.so
```
So the second one is obviously my fault, I can look into cleaning that properly (unless you are at it). The first seems to indicate that the test crashed somewhere?

File: src/ssh/Makefile-ssh.am [hunk 10]
```diff
@@ -81,8 +98,15 @@ EXTRA_DIST += \
...
 	src/ssh/mock-config \
 	src/ssh/invalid_known_hosts \
+	src/ssh/mock_known_hosts_2 \
 	$(NULL)
 
+if !HAVE_SSH_SESSION_HAS_KNOWN_HOSTS_ENTRY
+EXTRA_DIST += \
+	src/ssh/mock_known_hosts_2 \
+	$(NULL)
+endif
```
Comment (martinpitt) [inline src/ssh/Makefile-ssh.am]: This is duplicated and should be removed. EXTRA_DIST should never be conditionalized on configure tests, this defines what goes into our release tarballs.

File: src/ssh/cockpitsshknownhosts.c [hunk 11]
```diff
@@ -17,16 +17,147 @@
...
 #include <sys/types.h>
 #include <stdio.h>
 
-#include <glib/gstdio.h>
-#include "cockpitknownhosts.h"
+static gchar *knownhosts_file = NULL;
+
+void shim_set_knownhosts_file (const gchar *file)
+{
+  knownhosts_file = g_strdup (file);
```
Comment (martinpitt) [inline src/ssh/cockpitsshknownhosts.c]: This should `g_free (knownhosts_file)` first, in case this is being called multiple times.

File: src/ssh/cockpitsshrelay.c [hunk 12]
```diff
@@ -641,7 +604,7 @@ set_knownhosts_file (CockpitSshData *data,
    * a security check (if one can write /.ssh/known_hosts then we have to trust them), just caution */
   if (!host_known && g_strcmp0 (g_get_home_dir (), "/") != 0)
     {
-      host_known = cockpit_is_host_known (data->user_known_hosts, host, port);
+      host_known = session_has_known_host_in_file (data->user_known_hosts, data, host, port);
       if (host_known)
```
Comment (martinpitt) [inline src/ssh/cockpitsshrelay.c]: I really think/hope we can simplify `set_knownhosts_file()` a bit, but this can/should be a separate PR.

File: test/verify/check-multi-machine [hunk 13]
```diff
@@ -859,6 +859,12 @@ class TestMultiMachine(MachineCase):
         b.wait_present("#edit-machine-port")
         b.set_val("#edit-machine-port", "2222")
         b.click('#troubleshoot-dialog .btn-primary')
+        # Using the libssh's knownhosts api port is taken into account when verifying a host
```
Comment (martinpitt) [inline test/verify/check-multi-machine]: I just confirmed that ssh (CLI) behaves the same way, so this is really a bug fix. This exceptional case in the test is a bit ugly, but it will reduce over time. But would it be possible and reasonably easy to implement port checking in our shim as well? (follow-up is fine)

**Comment by martinpitt** [comment]
Grabbing this to make the following changes:
 * Move the dropping of legacy file test cases back into the proper commit
 * Reorder commits
 * Add unit test for the `[host]:port` change

**Comment by martinpitt** [comment]
Force-pushed with fixing the integration test change and ordering commits. I didn't add a unit test for the port change for now -- the integration tests already cover it, after all. This finally seems to work, great  work @Gundersanne!
